### PR TITLE
357 add variant with asset

### PR DESCRIPTION
**Pull request by heshamMassoud**
#### Summary
Added support for syncing assets of newly added variants.

Resolves issue #357  

#### Description
`ProductSyncUtils#buildActions` and `ProductUpdateActionUtils#buildVariantsUpdateActions` now build `AddAsset` actions for every new asset on every new variant on the new `ProductDraft`.
#### Todo

- Tests
    - [x] Unit 
    - [x] Integration
- [x] Documentation
- [x] Add Release Notes entry.

### CODE REVIEW

File: src/test/java/com/commercetools/sync/products/utils/ProductUpdateActionUtilsTest.java [hunk 1]
```diff
@@ -88,13 +96,35 @@ public void buildVariantsUpdateActions_updatesVariants() {
...
+        final ProductVariantDraft draftMaster = productDraftNew.getMasterVariant();
+        final ProductVariantDraft draft5 = productDraftNew.getVariants().get(1);
+        final ProductVariantDraft draft6 = productDraftNew.getVariants().get(2);
+        final ProductVariantDraft draft7 = productDraftNew.getVariants().get(3);
+
         assertThat(updateActions).contains(
-            buildAddVariantUpdateActionFromDraft(draftMaster),
-            buildAddVariantUpdateActionFromDraft(draft5),
-            buildAddVariantUpdateActionFromDraft(draft6));
+            AddVariant.of(draftMaster.getAttributes(), draftMaster.getPrices(), draftMaster.getSku(), true)
```
Comment (lojzatran) [inline src/test/java/com/commercetools/sync/products/utils/ProductUpdateActionUtilsTest.java]: Why this change? I don't see any assert with Assets here.

File: src/test/java/com/commercetools/sync/products/utils/ProductUpdateActionUtilsTest.java [hunk 2]
```diff
@@ -254,23 +284,106 @@ private void assertChangeMasterVariantEmptyErrorCatcher(final String productMock
...
+            assertThat(addVariant.getAttributes()).isSameAs(attributeList);
+            assertThat(addVariant.getPrices()).isSameAs(priceList);
+            assertThat(addVariant.getSku()).isEqualTo("testSKU");
+            assertThat(addVariant.getKey()).isEqualTo("testKey");
+            assertThat(addVariant.getImages()).isSameAs(imageList);
+        });
+    }
+
+    @Test
+    public void buildAddVariantUpdateActionFromDraft_WithNoAssets_BuildsNoAddAssets() {
```
Comment (lojzatran) [inline src/test/java/com/commercetools/sync/products/utils/ProductUpdateActionUtilsTest.java]: Isn't this test already included in `buildAddVariantUpdateActionFromDraft_WithAttribsPricesAndImages_ShouldBuildCorrectAddVariantAction`?

File: src/test/java/com/commercetools/sync/products/utils/ProductUpdateActionUtilsTest.java [hunk 3]
```diff
@@ -254,23 +284,106 @@ private void assertChangeMasterVariantEmptyErrorCatcher(final String productMock
...
+        final List<UpdateAction<Product>> result = buildAddVariantUpdateActionFromDraft(productVariantDraft);
+
+        // assertion
+        assertThat(result).containsExactlyInAnyOrder(
+            AddVariant.of(null, null, "foo", true)
+        );
+    }
+
+    @Test
+    public void buildAddVariantUpdateActionFromDraft_WithOneAsset_BuildsOneAddAssetsAction() {
```
Comment (lojzatran) [inline src/test/java/com/commercetools/sync/products/utils/ProductUpdateActionUtilsTest.java]: Isn't this test included in `buildAddVariantUpdateActionFromDraft_WithMultipleAsset_BuildsMultipleAddAssetsActions`?

File: src/test/java/com/commercetools/sync/products/utils/ProductUpdateActionUtilsTest.java (hunk 1 above)
Comment (heshamMassoud) [reply src/test/java/com/commercetools/sync/products/utils/ProductUpdateActionUtilsTest.java]: this is not related to assets. I changed to actually asserting if the correct actions are actually build  instead of using a method that is **under test** (`buildAddVariantUpdateActionFromDraft`) in the **test**..

File: src/test/java/com/commercetools/sync/products/utils/ProductUpdateActionUtilsTest.java (hunk 2 above)
Comment (heshamMassoud) [reply src/test/java/com/commercetools/sync/products/utils/ProductUpdateActionUtilsTest.java]: It's including but the assertions are different. The two tests are asserting different outcomes.

File: src/test/java/com/commercetools/sync/products/utils/ProductUpdateActionUtilsTest.java (hunk 3 above)
Comment (heshamMassoud) [reply src/test/java/com/commercetools/sync/products/utils/ProductUpdateActionUtilsTest.java]: Well, in this case, maybe it's kind of redundant.

Comment (heshamMassoud) [reply src/test/java/com/commercetools/sync/products/utils/ProductUpdateActionUtilsTest.java]: 343cc17

File: src/main/java/com/commercetools/sync/products/utils/ProductUpdateActionUtils.java [hunk 4]
```diff
@@ -595,13 +598,31 @@
...
+        final ArrayList<UpdateAction<Product>> actions = new ArrayList<>();
+
+        final UpdateAction<Product> addVariant = AddVariant
+            .of(draft.getAttributes(), draft.getPrices(), draft.getSku(), true)
+            .withKey(draft.getKey())
+            .withImages(draft.getImages());
+
+        actions.add(addVariant);
+
+        final List<UpdateAction<Product>> addAssetActions = ofNullable(draft.getAssets())
```
[the remark this answers is not in the document]
Comment (heshamMassoud) [reply src/main/java/com/commercetools/sync/products/utils/ProductUpdateActionUtils.java]: ops. Didn't notice that. Thanks: 5aa7775

File: src/test/java/com/commercetools/sync/products/utils/ProductUpdateActionUtilsTest.java (hunk 2 above)
Comment (lojzatran) [reply src/test/java/com/commercetools/sync/products/utils/ProductUpdateActionUtilsTest.java]: But assets are also one of the property, right? It's different action now because it's a bug, but it should not be treated as different property.

File: src/test/java/com/commercetools/sync/products/utils/ProductUpdateActionUtilsTest.java [hunk 5]
```diff
@@ -254,23 +284,83 @@ private void assertChangeMasterVariantEmptyErrorCatcher(final String productMock
...
+        // assertion
+        assertThat(result).containsExactlyInAnyOrder(
+            AddVariant.of(null, null, "foo", true)
+        );
+    }
+
+    @Test
+    public void buildAddVariantUpdateActionFromDraft_WithMultipleAssets_BuildsMultipleAddAssetsActions() {
+        // preparation
+        final AssetDraft assetDraft = AssetDraftBuilder
```
Comment (ahmetoz) [inline src/test/java/com/commercetools/sync/products/utils/ProductUpdateActionUtilsTest.java]: maybe IntStream can be used to not duplicate code:

```java
final List<AssetDraft> assetDrafts = IntStream.range(1, 4)
            .mapToObj(i-> AssetDraftBuilder
                .of(singletonList(AssetSourceBuilder.ofUri("foo").build()), ofEnglish("assetName"))
                .key(i + "")
                .build())
            .collect(toList());
```
